storage/badger.go

```diff
@@ -0,0 +1,230 @@
+package storage
+
+import (
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+	"log"
+	"os"
+)
+
+const (
+	// PerformanceMemTableSize is 3072 MB. Increases the maximum
+	// amount of data we can commit in a single transaction.
+	PerformanceMemTableSize = 3072 << 20
+
+	// PerformanceLogValueSize is 256 MB.
+	PerformanceLogValueSize = 256 << 20
+)
+
+type BadgerDatabase struct {
+	db            *badger.DB
+	opts          badger.Options
+	useWriteBatch bool
+}
+
+func NewBadgerDatabase(opts badger.Options, useWriteBatch bool) *BadgerDatabase {
+	return &BadgerDatabase{
+		db:            nil,
+		opts:          opts,
+		useWriteBatch: useWriteBatch,
+	}
+}
+
+func (bdb *BadgerDatabase) Setup() error {
+	db, err := badger.Open(bdb.opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+	bdb.db = db
+	return nil
+}
+
+func (bdb *BadgerDatabase) GetContext(id []byte) Context {
+	return NewBadgerContext(id, bdb.useWriteBatch)
+}
+
+func (bdb *BadgerDatabase) Update(ctx Context, fn func(Transaction, Context) error) error {
+	var wb *badger.WriteBatch
+	if bdb.useWriteBatch {
+		wb = bdb.db.NewWriteBatch()
+		defer wb.Cancel()
+		defer wb.Flush()
+	}
+
+	return bdb.db.Update(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, wb)
+		return fn(T, ctx)
+	})
+}
+
+func (bdb *BadgerDatabase) View(ctx Context, fn func(Transaction, Context) error) error {
+	return bdb.db.View(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, nil)
+		return fn(T, ctx)
+	})
+}
+
+func (bdb *BadgerDatabase) Close() error {
+	return bdb.db.Close()
+}
+
+func (bdb *BadgerDatabase) Erase() error {
+	return os.RemoveAll(bdb.opts.Dir)
+}
+
+func (bdb *BadgerDatabase) Id() DatabaseId {
+	return BADGERDB
+}
+
+// ==========================
+// BadgerTransaction
+// ==========================
+
+type BadgerTransaction struct {
+	txn *badger.Txn
+	wb  *badger.WriteBatch
+}
+
+func NewBadgerTransaction(txn *badger.Txn, wb *badger.WriteBatch) *BadgerTransaction {
+	return &BadgerTransaction{
+		txn: txn,
+		wb:  wb,
+	}
+}
+
+func (btx *BadgerTransaction) Set(key []byte, value []byte, ctx Context) error {
+	prefixedKey, err := castBadgerContextAndGetPrefixedKey(key, ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Set:")
+	}
+
+	if btx.wb != nil {
+		return btx.wb.Set(prefixedKey, value)
+	}
+	return btx.txn.Set(prefixedKey, value)
+}
+
+func (btx *BadgerTransaction) Delete(key []byte, ctx Context) error {
+	prefixedKey, err := castBadgerContextAndGetPrefixedKey(key, ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Delete:")
+	}
+
+	if btx.wb != nil {
+		return btx.wb.Delete(prefixedKey)
+	}
+	return btx.txn.Delete(prefixedKey)
+}
+
+func (btx *BadgerTransaction) Get(key []byte, ctx Context) ([]byte, error) {
```

Comment:
Ditto on removing the Context param here.
```suggestion
func (btx *BadgerTransaction) Get(key []byte) ([]byte, error) {
```